### DISCUSSION TITLE
Display PDF Page in PyQT5

### DISCUSSION THREAD

**Discussion by dmytro-balabukh** [pymupdf/PyMuPDF#1046] [Looking for help] [answered]
Im doing some work on my bookReaderProject. As I have mentioned in an article, I do it using PyQT5.
So the question is: What is the best way to represent pages in my Window, because now I'm trying to use HTML output, which I get from PyMuPDF, but the result is not as proper as I want to be. Another question is: can I decode dictionary, where pages are stored in blocks. Thanks

**Comment by JorjMcKie** [comment, marked as answer]
And why don't you use raster image output? I.e. use a pixmap and make a Qt image from it like so
```python
from PyQt<x>.QtGui import QImage
# ...
pix = page.get_pixmap(...)  # choose appropriate arguments,
# to control zooming and the right sub-rectangle part of the
# page
# set the correct QImage format depending on alpha
# ... presumably you want alpha=False
fmt = QImage.Format_RGBA8888 if pix.alpha else QImage.Format_RGB888
qtimg = QImage(pix.samples, pix.width, pix.height, pix.stride, fmt)
# use qtimg as a QImage
```

Of course nothing prevents you from using `page.get_text("dict")`.  You will get all text pieces including color, font, orientation, position and similarly images, so you can compose something within Qt with it.
For a high fidelity resemblance of the original page, drawings may be missing - which are often used in science publications instead of embedded images, to show things like Gantt charts, etc.
So you may need to also extract those elements in addition via `page.get_drawings()` and then use Qt's drawing capabilities to re-create them in your target. A complex undertaking, no doubt.
So you will most probably end up using bitmaps / pixmaps as shown above and focus your wits on including a nice zooming feature ...

**Reply by dmytro-balabukh** [reply]
Btw, I need to have a feature to highlight and copy text. That is why it is not the option

**Reply by JorjMcKie** [reply]
> Btw, I need to have a feature to highlight and copy text. That is why it is not the option

But I understood you want to show the PDF page image. This has nothing to do with **_also_** extracting / copying text or highlighting stuff.
You can (1) highlight text in the PDF page, (2) **_then_** make a pixmap of it - which pixmap now shows the highlight.
As long as you do not save the original PDF, all that is temporary.

**Reply by dmytro-balabukh** [reply]
Sorry, It's my fault that I didn't explain the whole situation. So I now Im on the ReaderProject.
My aim is to open PDF, and read it page by page. To do that I need to display those pages with all formatting, images and text.
It is like continuous access to the text. I did it with get_text("html") and loaded in PyQT5 QWebView, but as I mentioned it doesn't display properly. So I didn't understand the last think you suggested. And the last think - is it possible to get_text("json") or dict - really doesn't matter, edit this and load back in page. Thanks

**Reply by JorjMcKie** [reply]
I think I did understand.
As I said: the only reliable way to get a faithful (containing **_everything_**: text, images, drawings and what not) image of the rendered page is a pixmap.
This can easily be converted in a QImage as demonstrated, so Qt can display it.
In parallel, the text of that same page can be extracted, so you can work with it as you wish.

Now if you need to display the page with some modifications, like highlighted text, you simply do this highlighting **_before_** you render the page to a pixmap. In principle like this:
```python
page = doc[n]
search_rects = page.search_for("text to search for")
for rect in search_rects:
    page.addHighlightAnnot(rect)  # highlight the searched text
page = doc.reload_page(page)  # reload the page after adding highlights
pix = page.get_pixmap()
qimage = ...  # make a QImage of the page with the highlights
```
